matrix_functions.py

```diff
@@ -200,15 +203,37 @@ def matrix_eigendecomposition(
             device=A.device,
         )
 
+    # TODO: reduce redundant code when noninvertible_handling_config is generalized to all methods
+    # check epsilon is 0 when using pseudo-inverse
+    if (
+        isinstance(
+            eigendecomposition_config.noninvertible_handling_config,
+            PseudoInverseConfig,
+        )
+        and epsilon != 0.0
+    ):
+        raise ValueError(f"{epsilon=} should be 0.0 when using pseudo-inverse!")
+
+    # Add epsilon to the diagonal to help with numerical stability of the eigenvalue decomposition
+    # Only do it when add_epsilon_before_computation is True (root_inv_config must be a RegularizationConfig)
+    if getattr(eigendecomposition_config, "add_epsilon_before_computation", False):
+        dim = A.shape[0]
+        identity = torch.eye(dim, dtype=A.dtype, device=A.device)
+        A = A + epsilon * identity
+
+    # exclude noninvertible_handling_config from being passed to eigendecomposition method
+    kwargs = asdict(eigendecomposition_config)
+    kwargs.pop("noninvertible_handling_config", None)
```

Comment:
I would suggest to move this up and cache the popped config as a variable, i.e.:
```python
noninvertible_handling_config = kwargs.pop("noninvertible_handling_config", None)
```
Then this variable can be used instead of the long `eigendecomposition_config.noninvertible_handling_config`.